### PR TITLE
Handle `[UnscopedRef] scoped` in source and metadata

### PR DESCRIPTION
**Pull request by jcouv**
Fixes https://github.com/dotnet/roslyn/issues/63070 (metadata scenario)
Fixes https://github.com/dotnet/roslyn/issues/63057 (source scenario)

Relates to test plan https://github.com/dotnet/roslyn/issues/59194

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs [hunk 1]
```diff
@@ -13292,13 +13270,202 @@ public class A<T>
...
-                //         return ref F4A(ref r); // 3
-                Diagnostic(ErrorCode.ERR_EscapeCall, "F4A(ref r)").WithArguments("A<int>.F4A(ref R<int>)", "r4").WithLocation(21, 20),
-                // (21,28): error CS8168: Cannot return local 'r' by reference because it is not a ref local
-                //         return ref F4A(ref r); // 3
-                Diagnostic(ErrorCode.ERR_RefReturnLocal, "r").WithArguments("r").WithLocation(21, 28));
+                Diagnostic(ErrorCode.ERR_RefReturnLocal, "r").WithArguments("r").WithLocation(16, 28));
+        }
+
+        [Fact]
+        public void UnscopedScoped()
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:13277]: >UnscopedScoped [](http://example.com/codeflow?start=20&length=14)

This test looks like a duplicate of the previous test. If so, consider removing it.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs [hunk 2]
```diff
@@ -13292,13 +13270,202 @@ public class A<T>
...
+    }
+    public ref T F3A([UnscopedRef] ref R<T> r3)
+    {
+        return ref r3.F;
+    }
+}";
+            var comp = CreateCompilation(new[] { sourceA, UnscopedRefAttributeDefinition }, runtimeFeature: RuntimeFlag.ByRefFields);
+            comp.VerifyEmitDiagnostics();
+            var verifier = CompileAndVerify(comp, verify: Verification.Skipped);
+            var dump = verifier.Dump();
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:13304]: >var dump = verifier.Dump(); [](http://example.com/codeflow?start=12&length=27)

Is this needed?

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs [hunk 3]
```diff
@@ -13292,13 +13270,202 @@ public class A<T>
...
+    public ref T M([UnscopedRef] scoped out T t4) // 2
+    {
+        t4 = default;
+        return ref t4;
+    }
+    public ref T M2([UnscopedRef] scoped in R<T> t4) // 3
+    {
+        throw null;
+    }
+}";
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:13375]: >} [](http://example.com/codeflow?start=0&length=1)

Consider including `F([UnscopedRef] scoped R<T> r)` as well.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs [hunk 4]
```diff
@@ -13292,13 +13270,202 @@ public class A<T>
...
+            var sourceA =
+@"using System.Diagnostics.CodeAnalysis;
+public ref struct R<T>
+{
+    public ref T F;
+    public R(ref T t) { F = ref t; }
+}
+public class A<T>
+{
+    public ref T F4A([UnscopedRef] scoped ref R<T> r4) // 1
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:13362]: >F4A [](http://example.com/codeflow?start=17&length=3)

Minor: Consider renaming these F1, F2, F3, for consistency.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs [hunk 5]
```diff
@@ -202,7 +202,8 @@ internal sealed override DeclarationScope EffectiveScope
         {
             get
             {
-                var scope = DeclaredScope;
+                var scope = (RefKind == RefKind.Out && DeclaredScope == DeclarationScope.Unscoped) ? DeclarationScope.RefScoped : DeclaredScope;
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs:205]: Does it change behavior to move this logic from ParameterHelpers, etc. into this accessor? Or is it just a refactoring?

Comment (cston) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs:205]: The behavior change is described in `ParameterSymbol`. Basically, `DeclaredScope` is now just the declared scope, without the implicit scope for various cases. `EffectiveScope` is still the combined scope.

**Comment by cston** [comment]
@dotnet/roslyn-compiler, please review.

**Comment by cston** [comment]
dotnet/roslyn-compiler, please review.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs [hunk 6]
```diff
@@ -202,7 +202,7 @@ internal sealed override DeclarationScope EffectiveScope
         {
             get
             {
-                var scope = DeclaredScope;
+                var scope = ParameterHelpers.CalculateEffectiveScope(this);
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs:205]: nit: consider naming this something like *CalculateEffectiveScopeIgnoringAttributes*.

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs [hunk 7]
```diff
@@ -14668,26 +14651,143 @@ public class A<T>
...
+            VerifyParameterSymbol(type.GetMethod("F2").Parameters[0], "out T t2", RefKind.Out, DeclarationScope.Unscoped);
+            VerifyParameterSymbol(type.GetMethod("F3").Parameters[0], "in R<T> t3", RefKind.In, DeclarationScope.Unscoped);
+            VerifyParameterSymbol(type.GetMethod("F4").Parameters[0], "R<T> t4", RefKind.None, DeclarationScope.Unscoped);
+        }
+
+        [Fact, WorkItem(63070, "https://github.com/dotnet/roslyn/issues/63070")]
+        public void UnscopedScoped_Metadata()
+        {
+            var ilSource = """
+.class public sequential ansi sealed beforefieldinit R`1<T>
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:14721]: Please add a summary above of what this IL is equivalent to.

**Review by 333fred — APPROVED** [review]
LGTM (commit 5), with one test nit.

**Review by 333fred — APPROVED** [review]

**Pull request merged**